### PR TITLE
[pytest] Test the feature of monitoring critical processes by Supervisord.

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

### Description of PR

Summary:
This PR aims to test the [feature ](https://github.com/Azure/sonic-buildimage/pull/6242) of monitoring the critical processes by Supervisord.

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)

#### What is the motivation for this PR?
The motivation for this PR is to test the feature of monitoring the critical processes by Supervisord. This monitoring feature required Supervisord should write an alerting message into syslog if a critical process exited unexpectedly and the autorestart of container hosting this process was disabled.

#### How did you do it?
The logic of this PR is as follows:
Step 1: Records the autorestart status of all containers.
Step 2: Gets the number of ASICs on DUT.
Step 3: Gets all the containers which were enabled on DUT and at the same time for each **enabled** container, retrieve a list of namespaces which this container should be running.
Step 4: Disables the autorestart of all **enabled** containers.
Step 5: Manually stops all the critical process in all **enabled** containers on DUT.
Step 6: Waits for 70 seconds for the alerting messages to appear.
Step 7: Checks whether the expected alerting messages appeared in syslog or not.
Step 8: Executes the `config_reload(...)` to restart containers.
Step 9: Checks whether all critical processes in **enabled** containers are running.
Step 10: Restores the autorestart status of all containers.
Step 11: Does the post-check.

#### How did you verify/test it?
I tested this PR on a physical testbed with single ASIC.

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
N/A

### CODE REVIEW

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+        duthost: Hostname of DUT.
+        containers_in_namespaces: A dictionary where keys are container names and
+        values are lists which contains ids of namespaces this container should reside in.
+
+    Returns:
+        None.
+    """
+    logger.info("Checking the alerting messages from syslog...")
+    command_output = duthost.shell("sudo cat /var/log/syslog | grep '.*ERR.*supervisor-proc-exit-listener'",
+                                   module_ignore_errors=True)
```
Comment (lguohan) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:128]: why not use log analyzer to achieve this?

Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:128]: Will fix.

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+            for critical_group in critical_group_list:
+                group_program_info = get_group_program_info(duthost, container_name, critical_group)
+                for program_name in group_program_info:
+                    expected_alerting_messages.append(".*Process '{}' is not running in namespace '{}'.*".format(program_name, namespace_name))
+
+    logger.info("Generating the alerting message was done!")
+    return expected_alerting_messages
+
+
+def get_num_asics(duthost):
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:138]: Please use existing function: https://github.com/Azure/sonic-mgmt/blob/master/tests/common/devices.py#L638

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+        skip_containers: A list shows which containers will be skipped.
+
+    Returns:
+        A dictionary where keys are container names and values are a list which contains
+        ids of namespaces this container should reside in such as {lldp: ["host", "0", "1"]}
+    """
+    container_list = []
+    containers_in_namespaces = defaultdict(list)
+
+    container_list_command = "redis-cli -n 4 keys \"FEATURE|*\""
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:205]: please use sonic-db-cli instead.

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+    command_output = duthost.shell(container_list_command)
+    exit_code = command_output["rc"]
+    pytest_assert(exit_code == 0, "Failed to get keys (container names) in `FEATURE` table")
+    for line in command_output["stdout_lines"]:
+        container_list.append(line.split("|")[1].strip())
+
+    for container_name in container_list:
+        if container_name in skip_containers:
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:213]: why not remove the skip list from working list before the loop?

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 2 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:138]: Will update!

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+    marker = loganalyzer.init()
+
+    stop_critical_processes(duthost, containers_in_namespaces)
+
+    # Wait for 70 seconds such that Supervisord has a chance to write alerting message into syslog.
+    logger.info("Sleep 70 seconds to wait for the alerting message...")
+    time.sleep(70)
+
+    logger.info("Checking the alerting messages from syslog...")
+    loganalyzer.analyze(marker)
```
[the remark this answers is not in the document]
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:530]: Will update!

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 3 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:205]: Will update!

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+    for container_name in containers_in_namespaces.keys():
+        logger.info("Disabling the autorestart of container '{}'.".format(container_name))
+        command_disable_autorestart = "sudo config feature autorestart {} disabled".format(container_name)
+        command_output = duthost.shell(command_disable_autorestart)
+        exit_code = command_output["rc"]
+        pytest_assert(exit_code == 0, "Failed to disable the autorestart of container '{}'".format(container_name))
+        logger.info("The autorestart of container '{}' was disabled.".format(container_name))
+
+
+def get_group_program_info(duthost, container_name, group_name):
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:258]: Should this function be a shared helper function?

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+            group_program_info[program_name].append(program_pid)
+
+            if program_pid != -1:
+                logger.info("Found program '{}' in the '{}' state with pid {}"
+                            .format(program_name, program_status, program_pid))
+
+    return group_program_info
+
+
+def get_program_info(duthost, container_name, program_name):
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:297]: Should this function be a shared helper function?

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+
+                check_and_restart_process(duthost, container_name_in_namespace, critical_process)
+
+            for critical_group in critical_group_list:
+                group_program_info = get_group_program_info(duthost, container_name_in_namespace, critical_group)
+                for program_name in group_program_info:
+                    check_and_restart_process(duthost, container_name_in_namespace, program_name)
+
+
+def restore_containers_autorestart(duthost, containers_autorestart_states):
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:469]: You should combine this function with disable autostart function to be a fixture.

You should only reenable autorestart for contains you disabled earlier.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 6 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:258]: Moved this function to `tests/common/helers/dut_utils.py`.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 7 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:297]: Moved this function to `tests/common/helers/dut_utils.py`.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 8 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Added a fixture to disabled/enable the autorestart before/after the testing.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 4 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Fixed.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 2 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Fixed.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 3 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Fixed.

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+        duthost: Hostname of DUT.
+
+    Returns:
+        None.
+    """
+    containers_autorestart_states = duthost.get_container_autorestart_states()
+    disabled_autorestart_containers = []
+
+    for container_name, state in containers_autorestart_states.items():
+        if state != "disabled":
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:49]: Can you disable always_enabled feature?

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+    """Skips this test if the SONiC image installed on DUT was 201911 or old version.
+
+    Args:
+        duthost: Hostname of DUT.
+
+    Returns:
+        None.
+    """
+    if parse_version(duthost.kernel_version) <= parse_version("4.9.0"):
+        pytest.skip("Test was not supported for 201911 and older image version!")
```
Comment (yxieca) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:79]: You can use pytest_require to simplify the code structure. not a must fix.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 9 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py:49]: `always_enable` state is immutable and can not be changed.

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+        duthost: Hostname of DUT.
+
+    Returns:
+        None.
+    """
+    pytest_require(parse_version(duthost.kernel_version) > parse_version("4.9.0"),
+                   "Test was not supported for 201911 and older image versions!")
+
+
+def check_all_critical_processes_status(duthost):
```
Comment (jleveque) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:84]: Suggest renaming to `check_all_critical_processes_running`

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+    """Post-checks the status of critical processes.
```
Comment (jleveque) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:85]: This function could be called in more than a post-check scenario, if desired. Suggest rewording:

```
Determines if all critical processes are running on a DUT
```

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+
+            for critical_group in critical_group_list:
+                group_program_info = get_group_program_info(duthost, container_name, critical_group)
+                for program_name in group_program_info:
+                    check_and_kill_process(duthost, container_name_in_namespace, program_name,
+                                           group_program_info[program_name][0],
+                                           group_program_info[program_name][1])
+
+
+def check_and_restart_process(duthost, container_name, critical_process):
```
Comment (jleveque) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:291]: Suggest renaming: `ensure_process_is_running`

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+    """Checks the running status of a critical process and restarts it if it was not running.
```
Comment (jleveque) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:292]: If the process is stopped, we don't need to restart it. We should simply start it.

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+    else:
+        logger.info("Process '{}' in container '{}' is not running and restart it...".format(critical_process, container_name))
+        command_output = duthost.shell("docker exec {} supervisorctl restart {}".format(container_name, critical_process))
+        if command_output["rc"] == 0:
+            logger.info("Process '{}' in container '{}' is restarted.".format(critical_process, container_name))
+        else:
+            pytest.fail("Failed to restart process '{}' in container '{}'.".format(critical_process, container_name))
+
+
+def restart_critical_processes(duthost, containers_in_namespaces):
```
Comment (jleveque) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:315]: Again, this name doesn't accurately describe the function, because it only starts processes if they are not running. Suggest renaming: `ensure_all_critical_processes_running`

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+    """Checks whether all critical process are running.
```
Comment (jleveque) [inline tests/monitoring_processes/test_monitoring_critical_processes.py:316]: This also doesn't accurately describe the function. Please rewrite this.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 16 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Updated.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 15 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Great suggestion! Updated.

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 13 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Updated!

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 12 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Updated!

File: tests/monitoring_processes/test_monitoring_critical_processes.py (hunk 11 above)
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: Updated!

**Comment by lguohan** [comment]
check the lgtm alerts?

**Review by yxieca — APPROVED** [review]

File: tests/monitoring_processes/test_monitoring_critical_processes.py [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+            is_enabled = config_info[index + 1]
+        elif item == "has_global_scope":
+            has_global_scope = config_info[index + 1]
+        elif item == "has_per_asic_scope":
+            has_per_asic_scope = config_info[index + 1]
+
+    return is_enabled, has_global_scope, has_per_asic_scope
+
+
+def parse_feature_table(duthost, num_asics, skip_containers):
```
[the remark this answers is not in the document]
Comment (yozhao101) [reply tests/monitoring_processes/test_monitoring_critical_processes.py]: The function `parse_feature_table(...)` was used to get running containers list and which namespaces currently they are running. Currently this function was already removed and replaced by the function `get_containers_namespace_ids(...)` in which I used the `get_feature_status(...)` (https://github.com/Azure/sonic-mgmt/blob/67d5be22c740636c132047340d41fce163b9cee1/tests/common/devices.py#L972) to get the running container list.

File: tests/kvmtest.sh [hunk 18]
```diff
@@ -154,7 +155,8 @@ test_t1_lag() {
     lldp/test_lldp.py \
     route/test_default_route.py \
     platform_tests/test_cpu_memory_usage.py \
-    bgp/test_bgpmon.py"
+    bgp/test_bgpmon.py \
+    monitoring_processes/test_monitoring_critical_processes.py"
```
Comment (jleveque) [inline tests/kvmtest.sh:159]: Suggest renaming the directory to `process_monitoring` and the file to test_critical_process_monitoring.py

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]

File: tests/kvmtest.sh (hunk 18 above)
Comment (yozhao101) [reply tests/kvmtest.sh]: Updated!
